src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
+                            {
+                                await receiver.CompleteAsync(messagesArray.Select(x => x.SystemProperties.LockToken));
+                            }
+                        }
+                    }
+                    catch (ObjectDisposedException)
+                    {
+                        // Ignore as we are stopping the host
+                    }
+                    catch (NullReferenceException)
+                    {
+                        // Ignore as we are stopping the host
+                    }
+                    catch (Exception ex)
+                    {
+                        // Log another exception
+                        _logger.LogError($"Exception has occurred during batch listening:: {ex.ToString()}");
+                    }
+                }
+            }, cancellationToken);
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(null);
+            }
+        }
+
         public IScaleMonitor GetMonitor()
         {
             return _scaleMonitor.Value;
-        }
-    }
-}
+        }
+
+        private TriggeredFunctionData GetTriggerFunctionData(Message message, ServiceBusTriggerInput input)
+        {
+            Guid? parentId = ServiceBusCausalityHelper.GetOwner(message);
+            return new TriggeredFunctionData()
+            {
+                ParentId = parentId,
+                TriggerValue = input,
+                TriggerDetails = new Dictionary<string, string>()
+                {
+                    { "MessageId", message.MessageId },
+                    { "DeliveryCount", message.SystemProperties.DeliveryCount.ToString() },
+                    { "EnqueuedTime", message.SystemProperties.EnqueuedTimeUtc.ToString() },
+                    { "LockedUntil", message.SystemProperties.LockedUntilUtc.ToString() },
+                    { "SessionId", message.SessionId }
+                }
+            };
+        }
+
+        private TriggeredFunctionData GetTriggerFunctionData(Message[] messages, ServiceBusTriggerInput input)
+        {
+            Guid? parentId = ServiceBusCausalityHelper.GetOwner(messages[0]);
+
+            int length = messages.Length;
+            string[] messageIds = new string[length];
+            int[] deliveryCounts = new int[length];
+            DateTime[] enqueuedTimes = new DateTime[length];
+            DateTime[] lockedUntils = new DateTime[length];
+            string[] sessionIds = new string[length];
+            for (int i = 0; i < messages.Length; i++)
+            {
+                messageIds[i] = messages[i].MessageId;
+                deliveryCounts[i] = messages[i].SystemProperties.DeliveryCount;
+                enqueuedTimes[i] = messages[i].SystemProperties.EnqueuedTimeUtc;
+                lockedUntils[i] = messages[i].SystemProperties.LockedUntilUtc;
+                sessionIds[i] = messages[i].SessionId;

```

Comment:
Won't there still only be 1 session? Can a batch include messages from different sessions?

Reply:
You are right. There is only one session.